tutorial/lesson_12_using_the_gpu.cpp

```diff
@@ -283,37 +283,32 @@ int main(int argc, char **argv) {
 
 // A helper function to check if OpenCL, Metal or D3D12 is present on the host machine.
 
-#ifdef _WIN32
-#include <windows.h>
-#else
-#include <dlfcn.h>
-#endif
-
 Target find_gpu_target() {
     // Start with a target suitable for the machine you're running this on.
     Target target = get_host_target();
 
-    // Uncomment the following lines to try CUDA instead:
-    // target.set_feature(Target::CUDA);
-    // return target;
-
-#ifdef _WIN32
-    if (LoadLibraryA("d3d12.dll") != nullptr) {
-        target.set_feature(Target::D3D12Compute);
-    } else if (LoadLibraryA("OpenCL.dll") != nullptr) {
-        target.set_feature(Target::OpenCL);
-    }
-#elif __APPLE__
-    // OS X doesn't update its OpenCL drivers, so they tend to be broken.
-    // CUDA would also be a fine choice on machines with NVidia GPUs.
-    if (dlopen("/System/Library/Frameworks/Metal.framework/Versions/Current/Metal", RTLD_LAZY) != NULL) {
-        target.set_feature(Target::Metal);
+    std::vector<std::pair<DeviceAPI, Target::Feature>> devices_to_try;
```

Comment:
The parallel between DeviceAPI and Target::Feature is unfortunately boilerplatey. Is there a way to clean it up? Maybe it would be cleaner as host_supports_feature(Target::Feature), which dispatches internally to host_supports_device_api for features that correspond device APIs, and otherwise just calls get_host_target().has_feature(...);

Reply:
Yeah, we could do that, but I'm a little concerned that would invite misuse; host_supports_device_api() is clear that it is "special" (e.g. not threadsafe), whereas host_supports_feature() sounds more innocuous. 

Alternate/related would be to add something like `Target::required_device_api()`, which would return the DeviceAPI needed for a given featureset (or None if n/a); that would make the call much simpler.